test/integration/sds_generic_secret_integration_test.cc

```diff
@@ -159,4 +171,149 @@ TEST_P(SdsGenericSecretIntegrationTest, FilterFetchSuccess) {
+public:
+  SdsGenericSecretsPathConfigSourceTestFilterConfig() {
+    TestEnvironment::writeStringToFileForTest("encryption_key.txt", "DUMMY_AES_128_KEY");
+    TestEnvironment::writeStringToFileForTest("credential.txt", "PASSWORD");
+    auto secret = TestEnvironment::substitute(R"EOF(
+resources:
+- "@type": "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret"
+  name: encryption_key
+  generic_secret:
+    secrets:
+      encryption_key:
+        filename: "{{ test_tmpdir }}/encryption_key.txt"
+      credential:
+        filename: "{{ test_tmpdir }}/credential.txt"
+)EOF");
+    TestEnvironment::writeStringToFileForTest("generic_secret.yaml", secret);
+    auto secret_path = TestEnvironment::temporaryPath("generic_secret.yaml");
+    config_source_.mutable_path_config_source()->set_path(secret_path);
+  }
+};
+
+class SdsGenericSecretsPathConfigSourceIntegrationTest
+    : public Grpc::GrpcClientIntegrationParamTest,
+      public HttpIntegrationTest {
+public:
+  SdsGenericSecretsPathConfigSourceIntegrationTest()
+      : HttpIntegrationTest(Http::CodecType::HTTP1, ipVersion()), registration_(factory_) {}
+
+  void initialize() override {
+    config_helper_.prependFilter("{ name: sds-generic-secret-test }");
+    HttpIntegrationTest::initialize();
+  }
+
+  SdsGenericSecretsPathConfigSourceTestFilterConfig factory_;
+  Registry::InjectFactory<Server::Configuration::NamedHttpFilterConfigFactory> registration_;
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, SdsGenericSecretsPathConfigSourceIntegrationTest,
+                         GRPC_CLIENT_INTEGRATION_PARAMS);
+
+// This test verifies that multiple files specified in a SDS generic secret are watched and the
+// secret provider is up-to-date.
+TEST_P(SdsGenericSecretsPathConfigSourceIntegrationTest, GenericSecretFileUpdate) {
+  initialize();
+
+  codec_client_ = makeHttpConnection((lookupPort("http")));
+  Http::TestRequestHeaderMapImpl request_headers{
+      {":method", "GET"}, {":path", "/"}, {":scheme", "http"}, {":authority", "host"}};
+  sendRequestAndWaitForResponse(request_headers, 0, default_response_headers_, 0);
+
+  EXPECT_TRUE(upstream_request_->complete());
+  EXPECT_EQ(0U, upstream_request_->bodyLength());
+  EXPECT_EQ("DUMMY_AES_128_KEY", upstream_request_->headers()
+                                     .get(Http::LowerCaseString("encryption_key"))[0]
+                                     ->value()
+                                     .getStringView());
+  EXPECT_EQ("PASSWORD", upstream_request_->headers()
+                            .get(Http::LowerCaseString("credential"))[0]
+                            ->value()
+                            .getStringView());
+
+  // Update contents of all files specified in the secret.
+  TestEnvironment::writeStringToFileForTest("encryption_key.txt", "dummy_aes_128_key");
+  TestEnvironment::writeStringToFileForTest("credential.txt", "password");
```

Comment:
Can you clarify (in a comment) whether this invokes the `getDataSourceFilenames()` method once or twice?

Reply:
In my understanding, each file update triggers `getDataSourceFilenames()`.

Look at these functions:
1. Each file adds `onWatchUpdate` callback to the file system watcher: https://github.com/envoyproxy/envoy/blob/68f223448e9b6428f91d4ec8ded5170bb06683e0/source/common/secret/sds_api.cc#L133
2. `onWatchUpdate` calls `loadFiles`: https://github.com/envoyproxy/envoy/blob/68f223448e9b6428f91d4ec8ded5170bb06683e0/source/common/secret/sds_api.cc#L57
3. `loadFiles` calls `getDataSourceFilenames`: https://github.com/envoyproxy/envoy/blob/68f223448e9b6428f91d4ec8ded5170bb06683e0/source/common/secret/sds_api.cc#L211

So I guess this test triggers `getDataSourceFilenames()` at least 3 times.